chat_shell/chat_shell/tools/knowledge_factory.py

```
@@ -87,45 +97,54 @@ async def prepare_knowledge_base_tools(
-    kb_ls_tool._call_counter = exploration_call_counter
+    if restricted_search_only:
+        logger.info(
+            "[knowledge_factory] Created restricted KB tool set: knowledge_base_search only"
+        )
+    else:
+        # Create shared call counter for exploration tools (kb_ls and kb_head)
+        # These tools share the same max_calls_per_conversation limit as knowledge_base_search
+        # Get the actual limit from kb_tool configuration to ensure consistency
+        try:
+            max_calls, _ = kb_tool._get_kb_limits()
+        except Exception:
```

Ruff (0.15.6)

[warning] Do not catch blind exception: `Exception`

(BLE001)